### PR TITLE
Reapply " [XRay] Add support for instrumentation of DSOs on x86_64 (#90959)"

### PR DESCRIPTION
**Pull request by sebastiankreutzer**
This fixes remaining issues in my previous PR #90959.

Changes:
- Removed dependency on LLVM header in `xray_interface.cpp`
- Fixed XRay patching for some targets due to missing changes in architecture-specific patching functions
- Addressed some remaining compiler warnings that I missed in the previous patch
- Formatting

I have tested these changes on `x86_64` (natively), as well as `ppc64le`, `aarch64` and `arm32` (cross-compiled and emulated using qemu).

**Original description:**

This PR introduces shared library (DSO) support for XRay based on a revised version of the implementation outlined in [this RFC](https://discourse.llvm.org/t/rfc-upstreaming-dso-instrumentation-support-for-xray/73000).
The feature enables the patching and handling of events from DSOs, supporting both libraries linked at startup or explicitly loaded, e.g. via `dlopen`.
This patch adds the following:
- The `-fxray-shared` flag to enable the feature (turned off by default)
- A small runtime library that is linked into every instrumented DSO, providing position-independent trampolines and code to register with the main XRay runtime
- Changes to the XRay runtime to support management and patching of multiple objects

These changes are fully backward compatible, i.e. running without instrumented DSOs will produce identical traces (in terms of recorded function IDs) to the previous implementation.

Due to my limited ability to test on other architectures, this feature is only implemented and tested with x86_64. Extending support to other architectures is fairly straightforward, requiring only a position-independent implementation of the architecture-specific trampoline implementation (see `compiler-rt/lib/xray/xray_trampoline_x86_64.S` for reference). 

This patch does not include any functionality to resolve function IDs from DSOs for the provided logging/tracing modes. These modes still work and will record calls from DSOs, but symbol resolution for these functions in not available. Getting this to work properly requires recording information about the loaded DSOs and should IMO be discussed in a separate RFC, as there are mulitple feasible approaches.

### CODE REVIEW

File: compiler-rt/lib/xray/xray_interface.cpp [hunk 1]
```diff
@@ -150,21 +151,30 @@ class MProtectHelper {
...
+  return XRayInstrMaps[ObjId].Loaded;
+}
+
+bool patchSled(const XRaySledEntry &Sled, bool Enable, int32_t FuncId,
+               const XRayTrampolines &Trampolines) XRAY_NEVER_INSTRUMENT {
   bool Success = false;
   switch (Sled.Kind) {
   case XRayEntryType::ENTRY:
-    Success = patchFunctionEntry(Enable, FuncId, Sled, __xray_FunctionEntry);
+    Success = patchFunctionEntry(Enable, FuncId, Sled, Trampolines, false);
```
Comment (jplehr) [inline compiler-rt/lib/xray/xray_interface.cpp:168]:
```suggestion
    Success = patchFunctionEntry(Enable, FuncId, Sled, Trampolines, /* LogArgs=*/false);
```
There is some syntax like that, that is understood by clang-format and that helps to know what this is actually doing.

**Review by jplehr — APPROVED** [review]
Only a nit.
Let's see what happens.

**Comment by sebastiankreutzer** [comment]
For some reason, rebasing on main caused github to automatically request lots of reviews. Not sure why this is happening now.

**Comment by sebastiankreutzer** [comment]
Closing this PR, as something got mixed up during the rebase

**Pull request closed without merging**